test_runner/performance/test_sharded_ingest.py

```diff
@@ -0,0 +1,57 @@
+from __future__ import annotations
+
+from contextlib import closing
+
+import pytest
+from fixtures.benchmark_fixture import MetricReport, NeonBenchmarker
+from fixtures.common_types import Lsn, TenantShardId
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import (
+    NeonEnvBuilder,
+    wait_for_last_flush_lsn,
+)
+
+
+@pytest.mark.timeout(600)
+def test_sharded_ingest(neon_env_builder: NeonEnvBuilder, zenbenchmark: NeonBenchmarker):
+    """
+    Benchmarks sharded ingestion throughput, by ingesting a large amount of WAL into a Safekeeper
+    and fanning out to a large number of shards on dedicated Pageservers. Comparing the base case
+    (SHARD_COUNT=1) to the sharded case indicates the overhead of sharding.
+    """
+
+    # Specify the shard count and WAL volume. Each shard gets a dedicated pageserver.
+    SHARD_COUNT = 10
+    ROW_COUNT = 100_000_000  # about 7 GB of WAL
+
+    neon_env_builder.num_pageservers = SHARD_COUNT
+    env = neon_env_builder.init_start()
+
+    # Create a sharded tenant and timeline, and migrate it to the respective pageservers.
+    tenant_id, timeline_id = env.create_tenant(shard_count=SHARD_COUNT)
+
+    for shard_number in range(0, SHARD_COUNT):
+        tenant_shard_id = TenantShardId(tenant_id, shard_number, SHARD_COUNT)
+        pageserver_id = shard_number + 1
+        env.storage_controller.tenant_shard_migrate(tenant_shard_id, pageserver_id)
```

Comment:
This should happen on it's own, provided you call `storage_controller.reconcile_until_idle`. Still would be good to assert the following:
1. at the beginning of the test check that the shards are distributed correctly and stash the generation for each shard
2. at the end of the test, check that the generations are the same as in step 1 - this guarantees that no migrations have happened.

Reply:
Yeah, good call. For stuff like this I think it's better to explicitly specify the topology and disable any asynchronous actors, to ensure results are deterministic, but there doesn't seem to be a way to disable background reconciliation in the storage controller without also breaking foreground operations. I left a TODO to disable reconciliation and added assertions that shards don't move (the shard generation doesn't appear to be exposed via any HTTP APIs).